[mlir] Walk nested non-symbol table ops in symbol dce

jpienaar opened the pull request:
The previous code was effectively that a symbol is dead if was not nested in sequence of SymbolTables. But one can have operations that one cannot delete/DCE that refers to symbols which one could delete which resulted in symbol-dce deleting symbols that are still referenced and the resulting IR being invalid. This changes it so that all operations inside non SymbolTable op are considered to find nested SymbolTable ops.

j2kun commented on mlir/lib/Transforms/SymbolDCE.cpp:
```diff
@@ -22,6 +22,8 @@ namespace mlir {
 
 using namespace mlir;
 
+#define DEBUG_TYPE "symbol-dce"
```
IWYU nit: add `#include "llvm/Support/Debug.h"`

j2kun commented on mlir/lib/Transforms/SymbolDCE.cpp:
```diff
@@ -108,15 +112,42 @@ LogicalResult SymbolDCE::computeLiveness(Operation *symbolTableOp,
+      for (auto &region : op->getRegions()) {
+        for (auto &block : region.getBlocks()) {
+          for (Operation &op : block) {
+            SymbolOpInterface symbol = dyn_cast<SymbolOpInterface>(&op);
```
I'm mildly confused, but this is mostly for my education, not a request to change anything: isn't a symbol operation defined by the property that it "resides immediately within a region that defines a SymbolTable"? [[1]] If that is the case, it's not clear to me why this cast would ever trigger (the else block this path lies within implies the parent operation does not define a symbol table)

[1]: https://github.com/llvm/llvm-project/blob/9c52b3c52eb760a52826aad1368a4b0b46d924d3/mlir/include/mlir/IR/SymbolInterfaces.td#L26

j2kun replied:
Line 118

joker-eph replied:
Oh I got confused by the shadowing or using `op` in the loop line 138...
That's not great: we should use another name here.

jpienaar committed:
[mlir] Walk nested tables in symbol dce

The previous positioning was effectively that a symbol is dead if it cannot be addressed from top level. I think that is too strong a requirement: one can have operations that one cannot delete/DCE that refers to symbols which one could delete. This resulted in symbol-dce deleting symbols that are still referenced and the resulting IR being invalid.

This instead all the symbols of top level operations of non-symbol table ops additionally, as those are either dead and DCE would have handled, or alive and we cannot just delete symbols referenced internally. E.g., this treats non-symbol table regioned ops more conservatively.

jpienaar commented on mlir/lib/Transforms/SymbolDCE.cpp:
```diff
@@ -108,15 +112,42 @@ LogicalResult SymbolDCE::computeLiveness(Operation *symbolTableOp,
+          for (Operation &op : block) {
+            SymbolOpInterface symbol = dyn_cast<SymbolOpInterface>(&op);
+            if (!symbol) {
+              worklist.push_back(&op);
```
Updated to reflect that better, thanks

joker-eph commented on mlir/test/Transforms/test-symbol-dce.mlir:
```diff
@@ -98,3 +98,22 @@ module {
+  "test.one_region_op"() ({
+    "test.symbol_scope"() ({
+      // CHECK: func @nfunction
+      func.func @nfunction() {
```
```suggestion
      func.func nested @nfunction() {
```

I think that is one thing you're trying to fix.

However a simpler test right now would be:

```


module {
  "test.one_region_op"() ({
    module {
        func.func nested @nested() {
            return
        }

        func.func @main() {
            return
        }
    }
    "test.finish"() : () -> ()
  }) : () -> ()
}
```

Where @main should not be removed but @nested should be. Can you also add this test? It can be a good "documentation" as well of the expected behavior.

jpienaar replied:
Good idea, simplified to it (I think it actually shows it better, so no need to have both).

jpienaar committed:
Simplify test

joker-eph commented on mlir/lib/Transforms/SymbolDCE.cpp:
```diff
@@ -105,36 +110,69 @@ LogicalResult SymbolDCE::computeLiveness(Operation *symbolTableOp,
+      for (auto &region : op->getRegions())
+        for (auto &block : region.getBlocks())
+          for (Operation &op : block)
+            worklist.push_back(&op);
```
Shall we guard this behind `if (op->getNumRegions())` ?

```suggestion
            if (op->getNumRegions())
              worklist.push_back(&op);
```

jpienaar committed:
Update mlir/lib/Transforms/SymbolDCE.cpp

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

jpienaar committed:
Add test back and fix compile error

jpienaar committed:
Change whitespace in test to make more concise

jpienaar committed:
Update mlir/test/Transforms/test-symbol-dce.mlir

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

jpienaar committed:
Merge branch 'main' into symboldce
